src/Components/WebView/WebView/src/WebViewManager.cs

```diff
@@ -112,6 +112,63 @@ public Task AddRootComponentAsync(Type componentType, string selector, Parameter
+    /// <exception cref="ArgumentNullException">Thrown if <paramref name="workItem"/> is <c>null</c>.</exception>
+    /// <exception cref="InvalidOperationException">Thrown if this method is called before Blazor has started in the web view.</exception>
+    public async Task DispatchAsync(Action<IServiceProvider> workItem)
+    {
+        ArgumentNullException.ThrowIfNull(workItem);
+        if (_currentPageContext is null)
+        {
+            throw new InvalidOperationException($"{nameof(DispatchAsync)} can't be called until after Blazor has started in the web view.");
+        }
+
+        await _currentPageContext.Renderer.Dispatcher.InvokeAsync(() => workItem(_currentPageContext.ServiceProvider));
+    }
+
+    /// <summary>
+    /// Calls the specified <paramref name="workItem"/> asynchronously and passes in the scoped services available to Razor components.
+    /// </summary>
+    /// <typeparam name="TResult"></typeparam>
+    /// <param name="workItem">The action to call.</param>
+    /// <returns>A <see cref="Task"/> representing a value of type <typeparamref name="TResult"/> returned by the called <paramref name="workItem"/> that will be completed when the action has finished executing.</returns>
+    /// <exception cref="ArgumentNullException">Thrown if <paramref name="workItem"/> is <c>null</c>.</exception>
+    /// <exception cref="InvalidOperationException">Thrown if this method is called before Blazor has started in the web view.</exception>
+    public async Task<TResult> DispatchAsync<TResult>(Func<IServiceProvider, TResult> workItem)
+    {
+        ArgumentNullException.ThrowIfNull(workItem);
+        if (_currentPageContext is null)
+        {
+            throw new InvalidOperationException($"{nameof(DispatchAsync)} can't be called until after Blazor has started in the web view.");
+        }
+
+        return await _currentPageContext.Renderer.Dispatcher.InvokeAsync(() => workItem(_currentPageContext.ServiceProvider));
+    }
+
+    /// <summary>
+    /// Calls the specified <paramref name="workItem"/> asynchronously and passes in the scoped services available to Razor components.
+    /// </summary>
+    /// <param name="workItem">The action to call.</param>
+    /// <returns>Returns a <see cref="Task"/> representing <c>true</c> if the <paramref name="workItem"/> was called, or <c>false</c> if it was not called because Blazor is not currently running.</returns>
+    /// <exception cref="ArgumentNullException">Thrown if <paramref name="workItem"/> is <c>null</c>.</exception>
+    public async Task<bool> TryDispatchAsync(Action<IServiceProvider> workItem)
+    {
+        ArgumentNullException.ThrowIfNull(workItem);
+
+        if (_currentPageContext is null)
+        {
+            return false;
+        }
+
+        await _currentPageContext.Renderer.Dispatcher.InvokeAsync(() => workItem(_currentPageContext.ServiceProvider));
```

Comment:
Is it possible that, by the time the InvokeAsync callback runs, the page context has changed and this is no longer the current renderer? If that was the case, we'd be crossing the streams and running the callback inside the sync context for the old renderer, but supplying the service provider for the new renderer.

I don't 100% know if that could actually happen, but to avoid the risk, we could capture `_currentPageContext` in a local before L167 so that we know that even if it changes, we ensure consistency (the sync context and the service provider will always match, whether or not the page context has already changed).

Reply:
Update: actually I think there's slightly more to it still. Isn't it fun to think about race conditions? 🎉 

Simply ensuring consistency of sync context and service provider is not what we're trying to achieve here. That alone doesn't solve the original problem that, if the service provider you get is not current, then calling into it (e.g., JS interop) may fail. The goal is to ensure the service provider *is* current (and you're on the right sync context).

It looks to me like there isn't anything that locks `_currentPageContext`. It can change whenever the native webview reloads (which can happen any time - it's in a different process) and sends us an IPC message saying it did so. There may be a defect in `WebViewManager.AttachToPageAsync` in that it happily proceeds with disposing the old page context and creating a new one *without* using `_dispatcher` to synchronize anything. It would benefit from more careful analysis, but it looks like `MessageReceived` behaves better by using `_dispatcher`, and maybe `AttachToPageAsync` should do the same.

But even if we don't fix that, we could lean in more on the "try" part of `TryDispatchAsync`, and simply check whether or not the locally-captured copy of `_currentPageContext` still matches the actual-current one once we're inside the callback, and return `false` if not. Whether or not this is useful behavior still depends on exact use cases.